src/lib/FreeC/Pass/CompletePatternPass.hs

```diff
@@ -0,0 +1,123 @@





















































+--   The pattern matching is complete if there is exactly one case alternative





















































+--   for each constructor of the corresponding type.





















































+completePatternPass :: Pass IR.Module
+completePatternPass ast = do
+  mapM_ checkPatternFuncDecl (IR.modFuncDecls ast)
+  return ast
+
+-- | Checks a funcDecl for incomplete Patterns





















































+checkPatternFuncDecl :: IR.FuncDecl -> Converter ()
+checkPatternFuncDecl funcDecl = checkPatternExpr (IR.funcDeclRhs funcDecl)
+ where
+  checkPatternExpr :: IR.Expr -> Converter ()
+  checkPatternExpr (IR.Case srcSpan exprScrutinee exprAlts _) = do
+    checkPatternExpr exprScrutinee
+    mapM_ (checkPatternExpr . IR.altRhs) exprAlts
+    case IR.exprType (exprScrutinee) >>= getTypeConName of
+      Nothing       -> failedPatternCheck srcSpan
+      Just typeName -> do
+        env <- getEnv
+        let altConNames = map (IR.conPatName . IR.altConPat) exprAlts
+        case lookupEntry IR.TypeScope typeName env of
+          Just entry | isDataEntry entry ->
+            performCheck (entryConsNames entry) altConNames srcSpan
+          Just entry | isTypeSynEntry entry ->
+            case
+                getTypeConName (entryTypeSyn entry)
+                  >>= \e -> lookupEntry IR.TypeScope e env
+              of
+                Just entry' ->
+                  performCheck (entryConsNames entry') altConNames srcSpan
+                _ -> failedPatternCheck srcSpan
+          _ -> failedPatternCheck srcSpan
+  checkPatternExpr (IR.App _ lhr rhs _) =
+    checkPatternExpr lhr >> checkPatternExpr rhs
+  checkPatternExpr (IR.TypeAppExpr _ lhr _ _) = checkPatternExpr lhr
+  checkPatternExpr (IR.If _ exprCond exprThen exprElse _) =
+    checkPatternExpr exprCond
+      >> checkPatternExpr exprThen
+      >> checkPatternExpr exprElse
+  checkPatternExpr (IR.Lambda _ _ lambdaRhs _) = checkPatternExpr lambdaRhs
+  checkPatternExpr IR.Con{}                    = return ()
+  checkPatternExpr IR.Var{}                    = return ()
+  checkPatternExpr IR.Undefined{}              = return ()
+  checkPatternExpr IR.ErrorExpr{}              = return ()
+  checkPatternExpr IR.IntLiteral{}             = return ()
+
+  performCheck :: [IR.ConName] -> [IR.ConName] -> SrcSpan -> Converter ()
+  performCheck typeConNames altConNames srcSpan = unless
+    (  all (\x -> elem x typeConNames) typeConNames
+    && length typeConNames
+    == length altConNames
+    )
+    (failedPatternCheck srcSpan)
+
+  failedPatternCheck :: SrcSpan -> Converter ()
+  failedPatternCheck srcSpan =
+    reportFatal
+      $  Message srcSpan Error
+      $  "Incomplete pattern in function: "
+      ++ stringFromName (IR.funcDeclName funcDecl)
+
+  stringFromName :: IR.Name -> String
+  stringFromName (IR.Ident  s) = s
+  stringFromName (IR.Symbol s) = s
```

Comment:
Use [`showPretty` from `FreeC.Pretty`](https://freeproving.github.io/free-compiler/docs/master/FreeC-Pretty.html#v:showPretty) to embed IR AST nodes in error messages.

```suggestion
      ++ showPretty (IR.funcDeclName funcDecl)
```